libraries/shared/factories/event/repository_test.go

```diff
@@ -0,0 +1,211 @@
+			test_config.CleanTestDB(db)
+		})
+
+		BeforeEach(func() {
+			_, _ = db.Exec(createTestEventTableQuery)
+			_, _ = db.Exec(addCheckedColumnQuery)
+			headerRepository = repositories.NewHeaderRepository(db)
+			var insertHeaderErr error
+			headerID, insertHeaderErr = headerRepository.CreateOrUpdateHeader(fakes.FakeHeader)
+			Expect(insertHeaderErr).NotTo(HaveOccurred())
+			headerSyncLog := test_data.CreateTestLog(headerID, db)
+			logID = headerSyncLog.ID
+
+			testModel = event.InsertionModel{
+				SchemaName: "public",
+				TableName:  "testEvent",
+				OrderedColumns: []event.ColumnName{
+					event.HeaderFK, event.LogFK, "variable1",
+				},
+				ColumnValues: event.ColumnValues{
+					event.HeaderFK: headerID,
+					event.LogFK:    logID,
+					"variable1":    "value1",
+				},
+			}
+		})
+
+		AfterEach(func() {
+			db.MustExec(`DROP TABLE public.testEvent;`)
+			db.MustExec(`ALTER TABLE public.checked_headers DROP COLUMN testevent;`)
+		})
+
+		// Needs to run before the other tests, since those insert keys in map
+		It("memoizes queries", func() {
+			Expect(len(event.ModelToQuery)).To(Equal(0))
+			event.GetMemoizedQuery(testModel)
+			Expect(len(event.ModelToQuery)).To(Equal(1))
+			event.GetMemoizedQuery(testModel)
+			Expect(len(event.ModelToQuery)).To(Equal(1))
+		})
+
+		It("persists a model to postgres", func() {
+			createErr := event.Create([]event.InsertionModel{testModel}, db)
+			Expect(createErr).NotTo(HaveOccurred())
+
+			var res TestEvent
+			dbErr := db.Get(&res, `SELECT log_id, variable1 FROM public.testEvent;`)
+			Expect(dbErr).NotTo(HaveOccurred())
+
+			Expect(res.LogID).To(Equal(fmt.Sprint(testModel.ColumnValues[event.LogFK])))
+			Expect(res.Variable1).To(Equal(testModel.ColumnValues["variable1"]))
+		})
+
+		Describe("returns errors", func() {
+			It("for empty model slice", func() {
+				err := event.Create([]event.InsertionModel{}, db)
+				Expect(err).To(MatchError("repository got empty model slice"))
+			})
+
+			It("for failed SQL inserts", func() {
+				header := fakes.GetFakeHeader(1)
+				headerID, headerErr := headerRepository.CreateOrUpdateHeader(header)
+				Expect(headerErr).NotTo(HaveOccurred())
+
+				brokenModel := event.InsertionModel{
+					SchemaName: "public",
+					TableName:  "testEvent",
+					// Wrong name of last column compared to DB, will generate incorrect query
+					OrderedColumns: []event.ColumnName{
+						event.HeaderFK, event.LogFK, "variable2",
+					},
+					ColumnValues: event.ColumnValues{
+						event.HeaderFK: headerID,
+						event.LogFK:    logID,
+						"variable1":    "value1",
+					},
+				}
+
+				// Remove cached queries, or we won't generate a new (incorrect) one
+				delete(event.ModelToQuery, "publictestEvent")
+
+				createErr := event.Create([]event.InsertionModel{brokenModel}, db)
+				Expect(createErr).To(HaveOccurred())
+
+				// Remove incorrect query, so other tests won't get it
+				delete(event.ModelToQuery, "publictestEvent")
```

Comment:
Is there a way to handle this cleanup outside of the test? Thinking this won't happen if the assertion fails on line 141, and then there'll be a bunch of other unrelated failure dumped at the same time

Reply:
Hmm :thinking: Moving the deletion to before the assertion at least :)